source/opt/ccp_pass.cpp

```diff
@@ -139,19 +139,21 @@ SSAPropagator::PropStatus CCPPass::VisitAssignment(Instruction* instr) {
   Instruction* folded_inst =
       context()->get_instruction_folder().FoldInstructionToConstant(instr,
                                                                     map_func);
+
+  // Whether or not the instruction was folded, the folder may have created
+  // new constants.  If this happens, we need to indicate that CCP has modified
+  // the IR (independently of whether the new constant is actually propagated).
+  // See https://github.com/KhronosGroup/SPIRV-Tools/issues/3636 and
+  // https://github.com/KhronosGroup/SPIRV-Tools/issues/3991 for details.
+  if (context()->module()->IdBound() > next_id) {
```

Comment:
Does it make sense to pull this out to CCPPass::Process? Cache the id bound at the start of the pass and check it at the end?

Reply:
> Does it make sense to pull this out to CCPPass::Process? Cache the id bound at the start of the pass and check it at the end?

Yes, good observation.  It would simplify the tracking of whether constants were created.